pandas/io/json/_json.py

```diff
@@ -97,18 +102,19 @@ class Writer:
     def __init__(
         self,
         obj,
-        orient,
-        date_format,
-        double_precision,
-        ensure_ascii,
-        date_unit,
-        index,
-        default_handler=None,
+        orient: Optional[str],
+        date_format: str,
+        double_precision: int,
+        ensure_ascii: bool,
+        date_unit: str,
+        index: bool,
+        default_handler: Optional[Callable[[Any], Serializable]] = None,
+        indent: int = 0,
     ):
         self.obj = obj
 
         if orient is None:
-            orient = self._default_orient
+            orient = self._default_orient  # type: ignore
```

Comment:
`orient` is probably better as an abstract property but I figure best to leave that refactor to a follow up

Reply:
add a comment?

although adding an abstract property is straightforward,  if you add if TYPE-CHECKING it won't affect runtime behaviour.

```
if TYPE_CHECKING:
   _default_orient = None  # type: <...>
```

Reply:
no problem.

I don't think we should avoid these blocks as too beneficial.

in a (hopefully not too distant) future PR, i'll be adding...

```
    if TYPE_CHECKING:  # attributes index and columns are created dynamically
        index = None  # type: Index
        columns = None  # type: Index
```

to frame.py so that i can get types for `self.fmt.tr_frame.index.format` and `self.frame.index.nlevels` etc.

in this case all tests pass without the `if TYPE_CHECKING` but would rather not change runtime behaviour if avoidable.